### PR TITLE
MTDSA-3145 - Create Crystallisation V2 request parsers and validators

### PR DESCRIPTION
**Pull request by jeremystone**

### CODE REVIEW

File: app/v2/controllers/requestParsers/validators/validations/CalculationIdValidation.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package v2.controllers.requestParsers.validators.validations
+
+import v2.models.errors
+import v2.models.errors.InvalidCalcIdError
+
+object CalculationIdValidation extends RegexValidation {
+  override val regexFormat: String = "^[0-9]{8}$|^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
+
+  override def error: errors.Error = InvalidCalcIdError
```
Comment (ma3574) [inline app/v2/controllers/requestParsers/validators/validations/CalculationIdValidation.scala]: Would it be easier to import `errors.Error` at the top of the file and not have `error: errors.Error`

Comment (ma3574) [reply app/v2/controllers/requestParsers/validators/validations/CalculationIdValidation.scala]: `val`

File: app/v2/controllers/requestParsers/validators/validations/RegexValidation.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package v2.controllers.requestParsers.validators.validations
+
+import v2.models.errors.Error
+
+trait RegexValidation {
+  def regexFormat: String
+
+  def error: Error
```
Comment (ma3574) [inline app/v2/controllers/requestParsers/validators/validations/RegexValidation.scala]: `val`

File: app/v2/controllers/requestParsers/validators/validations/RegexValidation.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2.controllers.requestParsers.validators.validations
+
+import v2.models.errors.Error
+
+trait RegexValidation {
+  def regexFormat: String
```
Comment (ma3574) [inline app/v2/controllers/requestParsers/validators/validations/RegexValidation.scala]: `val`

File: app/v2/models/errors/mtdErrors.scala [hunk 4]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
+object RuleIncorrectOrEmptyBodyError extends Error("RULE_INCORRECT_OR_EMPTY_BODY_SUBMITTED", "An empty or non-matching body was submitted")
 object IncomeSourcesChangedError extends Error("RULE_INCOME_SOURCES_CHANGED", "The remote endpoint has indicated changed information for income sources.  Recalculate before crystallising")
 object RecentSubmissionsExistError extends Error("RULE_RECENT_SUBMISSIONS_EXIST", "The remote endpoint has indicated more recent submissions exist.  Recalculate before crystallising")
 object ResidencyChangedError extends Error("RULE_RESIDENCY_CHANGED", "The remote endpoint has indicated a change in residency.  Recalculate before crystallising")
 object FinalDeclarationReceivedError extends Error("RULE_FINAL_DECLARATION_RECEIVED", "The remote endpoint has indicated that final declaration has already been received")
+
+//Standard Errors
+object NotFoundError extends Error("MATCHING_RESOURCE_NOT_FOUND", "The remote endpoint has indicated that no intent-to-crystallise calculation exists for the calculationId")
```
Comment (ma3574) [inline app/v2/models/errors/mtdErrors.scala]: Need to check what @narendravijayarao has added to the RAML as this text will need to match that.

File: test/v2/controllers/requestParsers/CrystallisationRequestDataParserSpec.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import play.api.mvc.AnyContentAsJson
+import support.UnitSpec
+import uk.gov.hmrc.domain.Nino
+import v2.mocks.validators.MockCrystallisationValidator
+import v2.models.domain.CrystallisationRequest
+import v2.models.errors._
+import v2.models.requestData.{CrystallisationRawData, CrystallisationRequestData, DesTaxYear}
+
+/**
+  * Tests for [[CrystallisationRequestDataParser]]
```
Comment (ma3574) [inline test/v2/controllers/requestParsers/CrystallisationRequestDataParserSpec.scala]: Let's remove

File: test/v2/controllers/requestParsers/validators/CrystallisationValidatorSpec.scala [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+package v2.controllers.requestParsers.validators
+
+import play.api.libs.json.Json
+import play.api.mvc.AnyContentAsJson
+import support.UnitSpec
+import v2.models.errors._
+import v2.models.requestData.CrystallisationRawData
+
+/**
+  * Tests for [[CrystallisationValidator]]
```
Comment (ma3574) [inline test/v2/controllers/requestParsers/validators/CrystallisationValidatorSpec.scala]: Let's remove

File: test/v2/controllers/requestParsers/validators/CrystallisationValidatorSpec.scala [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+          List(TaxYearFormatError)
+      }
+    }
+
+    "return RuleTaxYearNotSupportedError error" when {
+      "an out of range tax year is supplied" in {
+                        validator.validate(
+                          CrystallisationRawData(validNino, "2016-17",  body(validJson))) shouldBe
+                            List(RuleTaxYearNotSupportedError)
+        fail("is this validation reqd - remove validation if not?")
```
Comment (ma3574) [inline test/v2/controllers/requestParsers/validators/CrystallisationValidatorSpec.scala]: 👍

File: test/v2/controllers/requestParsers/validators/validations/CalculationIdValidationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package v2.controllers.requestParsers.validators.validations
+
+import support.UnitSpec
+import v2.models.errors.InvalidCalcIdError
+
+/**
+  * Tests for [[CalculationIdValidation]]
```
Comment (ma3574) [inline test/v2/controllers/requestParsers/validators/validations/CalculationIdValidationSpec.scala]: Let's remove

File: app/v2/controllers/requestParsers/validators/CrystallisationValidator.scala [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    )
+  }
+
+  private def bodyFormatValidator: CrystallisationRawData => List[List[Error]] = { data =>
+    List(
+      JsonFormatValidation.validate[CrystallisationRequest](data.body, RuleIncorrectOrEmptyBodyError)
+    )
+  }
+
+  private def bodyFieldsValidation: CrystallisationRawData => List[List[Error]] = (data: CrystallisationRawData) => {
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/CrystallisationValidator.scala]: Dangerously close to the forbidden purple syntax... Is there any reason to write a function like this over `private def bodyFieldsValidation(data: CrystallisationRawData): List[List[Error]] = { ... }`?

File: app/v2/models/errors/mtdErrors.scala [hunk 10]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
 object NinoFormatError extends Error("FORMAT_NINO", "The NINO format is invalid")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Spec reads "The provided NINO is invalid"

File: app/v2/models/errors/mtdErrors.scala [hunk 11]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
 
 // MTD Errors
-object InvalidTaxYearError extends Error("FORMAT_TAX_YEAR", "The provided tax year is invalid")
+object TaxYearFormatError extends Error("FORMAT_TAX_YEAR", "The provided tax year is invalid")
 object InvalidCalcIdError extends Error("FORMAT_CALCID", "The provided calculationId is invalid")
 
-// Backend Response Errors
-object NotFoundError extends Error("MATCHING_RESOURCE_NOT_FOUND", "The remote endpoint has indicated that no intent-to-crystallise calculation exists for the calculationId")
+// Rule Errors
+object RuleTaxYearNotSupportedError extends Error("RULE_TAX_YEAR_NOT_SUPPORTED", "Tax year not supported, because it precedes the earliest allowable tax year")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Not in the Spec - where is this from?

File: app/v2/models/errors/mtdErrors.scala [hunk 12]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
+object RuleIncorrectOrEmptyBodyError extends Error("RULE_INCORRECT_OR_EMPTY_BODY_SUBMITTED", "An empty or non-matching body was submitted")
 object IncomeSourcesChangedError extends Error("RULE_INCOME_SOURCES_CHANGED", "The remote endpoint has indicated changed information for income sources.  Recalculate before crystallising")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Spec reads "Income sources data has changed. Perform intent to crystallise"

File: app/v2/models/errors/mtdErrors.scala [hunk 13]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
 object RecentSubmissionsExistError extends Error("RULE_RECENT_SUBMISSIONS_EXIST", "The remote endpoint has indicated more recent submissions exist.  Recalculate before crystallising")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Spec reads "More recent submissions exist. Perform intent to crystallise"

File: app/v2/models/errors/mtdErrors.scala [hunk 14]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
 object ResidencyChangedError extends Error("RULE_RESIDENCY_CHANGED", "The remote endpoint has indicated a change in residency.  Recalculate before crystallising")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Spec reads "Residency has changed.  Perform intent to crystallise"

File: app/v2/models/errors/mtdErrors.scala [hunk 15]
```diff
@@ -20,15 +20,19 @@ package v2.models.errors
...
 object FinalDeclarationReceivedError extends Error("RULE_FINAL_DECLARATION_RECEIVED", "The remote endpoint has indicated that final declaration has already been received")
```
Comment (james-work-account) [inline app/v2/models/errors/mtdErrors.scala]: Spec reads "Crystallisation declaration has already been received"

File: app/v2/models/errors/mtdErrors.scala (hunk 11 above)
Comment (jeremystone) [reply app/v2/models/errors/mtdErrors.scala]: It was yesterday. We had a long conversation about it. Strange!

File: app/v2/controllers/requestParsers/validators/CrystallisationValidator.scala (hunk 9 above)
Comment (jeremystone) [reply app/v2/controllers/requestParsers/validators/CrystallisationValidator.scala]: Nice try but these do need to be functions. They are added to a collection of functions : validationSet.

Comment (james-work-account) [reply app/v2/controllers/requestParsers/validators/CrystallisationValidator.scala]: :suspect:

**Pull request merged**